### ISSUE TITLE
[Metrics] Model Container Metric Server Design

### ISSUE THREAD

**Issue by simon-mo** [ucbrise/clipper#423]
## Model Container Metric Server Design

To implement #412 and #401, we planed to build a small metric server that was spawned by `rpc.py` and exposes the api to accept any user defined metrics or update default configuration. 

After experimenting with a small metric server inside model container RPC, the performance is not optimal. For a small python flask server accepting metrics dictionary via http post call. Each metrics take around 4-6ms. The latency is high given that users will most likely report metrics during prediction process. 

An alternative is use redis as intermediate message queue, it has following advantage: 

- Small and lightweight: we can just install redis binary inside model container. 
- Highly performant: using `redis-py` client, user can dump the metrics dictionary in 140 µs.
- Language agnostic: almost every language has a redis client library. Therefore, models in Java and R can also report metrics if we adopt such schemes. 

The detail is design is as follows:

- `metric_server.py`: a self-contained metric server script that 

  1. Collect message from redis via a pub/sub channel 
  2. Process message: add metric, record metric, update rules according to some documented api. 
  3. Expose metrics for prometheus to scrape. 

- `rpc.py` will call `metric_server.py` as a subprocess. Other containers like R-container, jvm based container can just call this script as subprocess as well. 

- Inside user's model, if the user wants to report metric, we can provide utils library to abstract away the redis pattern:

  ```python
  from clipper_admin.metric_utils import add_udf, report_udf

  add_udf('logistic regression class prob', 'Histogram')
  while True:
  	report_udf('Class Probability', 0.87)
      ...
  ```

**Comment by simon-mo** [comment]
@Corey-Zumar any comments?

**Comment by Corey-Zumar** [comment]
@Simon-Mo This makes sense. Can you elaborate on the process of communicating between the metrics-server subprocess and the rpc process? Metrics data from a model class inheriting from `rpc.ModelContainerBase` will need to be transmitted to the metrics subprocess and then persisted to Redis via the pub-sub mechanism. Do you intend to use a multi-process queue or some other piece of MP infrastructure to accomplish this?

**Comment by Corey-Zumar** [comment]
@Simon-Mo Ah, I see. Your proposed multi-process communication strategy is the use of the Redis channel, right? We'll want to make sure that the RPC implementation makes calls to the Redis channel on a separate thread; we don't want metrics message passing overheads to detract from RPC performance; this should be straightforward in R and C++, but the GIL may complicate the task in Python.

**Comment by simon-mo** [comment]
@Corey-Zumar It is true the redis is now a multi-process communication strategy. 

Addressing other points:
- I ran some numbers: from a raw ({str: int, str: float}) dictionary via pickle to send out the metric:
   1. redis-py takes 85 µs over TCP, 42 µs over unix socket.
   2. zeromq takes 80 µs
   2. [credis](https://github.com/yihuang/credis) takes 53 µs over TCP, **30 µs over unix socket**.
 (redis binding implemented in cython)
   2. multiprocessing.pipe takes 18 µs (buffer is limited)
   3. multiprocessing.queue takes 12 µs (buffer is limited)
- The benefit of redis is multi-language client support. The downside is we need to take the overhead into account. Multi-threading doesn't really work in python due to GIL. 
- The general idea is:
   1. The metric reporting interface should be a unified interface, add/modify/report metrics should be done through some message passing interface. 
  2. A separate python metric server will handle these message and prepare metrics for prometheus.

**Comment by simon-mo** [comment]
Closed via #440
